Description: Adapt KUserFeedback 6.30 PHP autotests to PHPUnit 13 and scoped test loading
Author: SupraLINUX Project <packages@supralinux.invalid>
Forwarded: not-yet
Last-Update: 2026-09-20
Origin: SupraLINUX
Bug: https://invent.kde.org/frameworks/kuserfeedback
---
diff --git a/autotests/utilstest.php b/autotests/utilstest.php
--- a/autotests/utilstest.php
+++ b/autotests/utilstest.php
@@ -1,81 +1,81 @@
-<?php
-/*
-    SPDX-FileCopyrightText: 2016 Volker Krause <vkrause@kde.org>
-
-    SPDX-License-Identifier: MIT
-*/
-
-require_once('../src/server/shared/utils.php');
-
-class UtilsTest extends PHPUnit\Framework\TestCase
-{
-    public static function isValidIdentifier_data()
-    {
-        return [
-            'empty' => [ '', false ],
-            'number' => [ 42, false ],
-            'numstring' => [ '42', false ],
-            'leading number' => [ '1foo', false ],
-            'alpha only' => [ 'foo', true ],
-            'leading space' => [ ' foo', false ],
-            'middle space' => [ 'f o o', false ],
-            'trailing space' => [ 'foo ', false ],
-            'valid' => [ 'foo42', true ],
-            'leading underscore' => [ '_foo', true ],
-            'underscore' => [ 'f_o_o', true ],
-            'dots and dashes' => [ 'org.kde.unit-test', true ],
-            'control' => [ "fo\no", false ]
-        ];
-    }
-
-    /** @dataProvider isValidIdentifier_data */
-    public function testIsValidIdentifier($str, $result)
-    {
-        $this->assertEquals($result, Utils::isValidIdentifier($str));
-    }
-
-    public static function normalize_data()
-    {
-        return [
-            'empty' => [ '', '' ],
-            'normal' => [ 'foo', 'foo' ],
-            'dot' => [ 'org.kde.foo', 'org_kde_foo' ]
-        ];
-    }
-
-    /** @dataProvider normalize_data */
-    public function testNormalize($input, $output)
-    {
-        $this->assertEquals($output, Utils::normalizeString($input));
-    }
-
-    public static function primaryKeyColumn_data()
-    {
-        return [
-            'sqlite' => [ 'sqlite', 'id', 'id INTEGER PRIMARY KEY AUTOINCREMENT' ],
-            'pgsql' => [ 'pgsql', 'id', 'id SERIAL PRIMARY KEY' ],
-            'mysql' => [ 'mysql', 'id', 'id INTEGER PRIMARY KEY AUTO_INCREMENT' ]
-        ];
-    }
-
-    /** @dataProvider primaryKeyColumn_data */
-    public function primaryKeyColumn($driver, $name, $output)
-    {
-        $this->assertEquals($output, Utils::primaryKeyColumnDeclaration($driver, $name));
-    }
-
-    public static function sqlStringType_data()
-    {
-        return [
-            'sqlite' => [ 'sqlite', 'VARCHAR' ],
-            'mysql' => [ 'mysql', 'VARCHAR(255)' ],
-            'pgsql' => [ 'pgsql', 'VARCHAR' ],
-        ];
-    }
-
-    /** @dataProvider sqlStringType_data */
-    public function testSqlStringType($driver, $output)
-    {
-        $this->assertEquals($output, Utils::sqlStringType($driver));
-    }
-}
+<?php
+/*
+    SPDX-FileCopyrightText: 2016 Volker Krause <vkrause@kde.org>
+
+    SPDX-License-Identifier: MIT
+*/
+
+require_once('../src/server/shared/utils.php');
+
+class UtilsTest extends PHPUnit\Framework\TestCase
+{
+    public static function isValidIdentifier_data()
+    {
+        return [
+            'empty' => [ '', false ],
+            'number' => [ 42, false ],
+            'numstring' => [ '42', false ],
+            'leading number' => [ '1foo', false ],
+            'alpha only' => [ 'foo', true ],
+            'leading space' => [ ' foo', false ],
+            'middle space' => [ 'f o o', false ],
+            'trailing space' => [ 'foo ', false ],
+            'valid' => [ 'foo42', true ],
+            'leading underscore' => [ '_foo', true ],
+            'underscore' => [ 'f_o_o', true ],
+            'dots and dashes' => [ 'org.kde.unit-test', true ],
+            'control' => [ "fo\no", false ]
+        ];
+    }
+
+    #[\PHPUnit\Framework\Attributes\DataProvider('isValidIdentifier_data')]
+    public function testIsValidIdentifier($str, $result)
+    {
+        $this->assertEquals($result, Utils::isValidIdentifier($str));
+    }
+
+    public static function normalize_data()
+    {
+        return [
+            'empty' => [ '', '' ],
+            'normal' => [ 'foo', 'foo' ],
+            'dot' => [ 'org.kde.foo', 'org_kde_foo' ]
+        ];
+    }
+
+    #[\PHPUnit\Framework\Attributes\DataProvider('normalize_data')]
+    public function testNormalize($input, $output)
+    {
+        $this->assertEquals($output, Utils::normalizeString($input));
+    }
+
+    public static function primaryKeyColumn_data()
+    {
+        return [
+            'sqlite' => [ 'sqlite', 'id', 'id INTEGER PRIMARY KEY AUTOINCREMENT' ],
+            'pgsql' => [ 'pgsql', 'id', 'id SERIAL PRIMARY KEY' ],
+            'mysql' => [ 'mysql', 'id', 'id INTEGER PRIMARY KEY AUTO_INCREMENT' ]
+        ];
+    }
+
+    #[\PHPUnit\Framework\Attributes\DataProvider('primaryKeyColumn_data')]
+    public function primaryKeyColumn($driver, $name, $output)
+    {
+        $this->assertEquals($output, Utils::primaryKeyColumnDeclaration($driver, $name));
+    }
+
+    public static function sqlStringType_data()
+    {
+        return [
+            'sqlite' => [ 'sqlite', 'VARCHAR' ],
+            'mysql' => [ 'mysql', 'VARCHAR(255)' ],
+            'pgsql' => [ 'pgsql', 'VARCHAR' ],
+        ];
+    }
+
+    #[\PHPUnit\Framework\Attributes\DataProvider('sqlStringType_data')]
+    public function testSqlStringType($driver, $output)
+    {
+        $this->assertEquals($output, Utils::sqlStringType($driver));
+    }
+}
diff --git a/autotests/schemaentrytest.php b/autotests/schemaentrytest.php
--- a/autotests/schemaentrytest.php
+++ b/autotests/schemaentrytest.php
@@ -1,55 +1,53 @@
-<?php
-/*
-    SPDX-FileCopyrightText: 2016 Volker Krause <vkrause@kde.org>
-
-    SPDX-License-Identifier: MIT
-*/
-
-require_once('../src/server/shared/product.php');
-require_once('../src/server/shared/schemaentry.php');
-
-class SchemaEntryTest extends PHPUnit\Framework\TestCase
-{
-    public function dataTableName_data()
-    {
-        return [
-            'normal' => [ 'foo', 'pd2_org_kde_testproduct__foo' ],
-            'dot' => [ 'my.value', 'pd2_org_kde_testproduct__my_value' ],
-            'uppercase' => [ 'UPPER', 'pd2_org_kde_testproduct__upper' ]
-        ];
-    }
-
-    /** @dataProvider dataTableName_data */
-    public function testDataTableName($input, $output)
-    {
-        $p = new Product;
-        $p->name = 'org.kde.TestProduct';
-        $se = new SchemaEntry($p);
-        $se->name = $input;
-        $this->assertEquals($output, $se->dataTableName());
-    }
-
-    public function invalidJson_data()
-    {
-        return [
-            'empty' => [ '{}' ],
-            'empty name' => [ '{ "name": "", "type": "scalar", "aggregationType": "none" }' ],
-            'empty type' => [ '{ "name": "foo", "type": "", "aggregationType": "none" }' ],
-            'invalid type' => [ '{ "name": "foo", "type": "bla", "aggregationType": "none" }' ],
-            'invalid name' => [ '{ "name": " foo ", "type": "scalar", "aggregationType": "none" }' ],
-            'invalid name 2' => [ '{ "name": "1foo ", "type": "scalar", "aggregationType": "none" }' ]
-        ];
-    }
-
-    /**
-     * @dataProvider invalidJson_data
-     */
-    public function testInvalidJson($input)
-    {
-        $this->expectException(RESTException::class);
-        $this->expectExceptionCode(400);
-        $p = new Product;
-        $se = SchemaEntry::fromJson(json_decode('[ ' . $input . ' ]'), $p);
-    }
-}
-
+<?php
+/*
+    SPDX-FileCopyrightText: 2016 Volker Krause <vkrause@kde.org>
+
+    SPDX-License-Identifier: MIT
+*/
+
+require_once('../src/server/shared/product.php');
+require_once('../src/server/shared/schemaentry.php');
+
+class SchemaEntryTest extends PHPUnit\Framework\TestCase
+{
+    public static function dataTableName_data()
+    {
+        return [
+            'normal' => [ 'foo', 'pd2_org_kde_testproduct__foo' ],
+            'dot' => [ 'my.value', 'pd2_org_kde_testproduct__my_value' ],
+            'uppercase' => [ 'UPPER', 'pd2_org_kde_testproduct__upper' ]
+        ];
+    }
+
+    #[\PHPUnit\Framework\Attributes\DataProvider('dataTableName_data')]
+    public function testDataTableName($input, $output)
+    {
+        $p = new Product;
+        $p->name = 'org.kde.TestProduct';
+        $se = new SchemaEntry($p);
+        $se->name = $input;
+        $this->assertEquals($output, $se->dataTableName());
+    }
+
+    public static function invalidJson_data()
+    {
+        return [
+            'empty' => [ '{}' ],
+            'empty name' => [ '{ "name": "", "type": "scalar", "aggregationType": "none" }' ],
+            'empty type' => [ '{ "name": "foo", "type": "", "aggregationType": "none" }' ],
+            'invalid type' => [ '{ "name": "foo", "type": "bla", "aggregationType": "none" }' ],
+            'invalid name' => [ '{ "name": " foo ", "type": "scalar", "aggregationType": "none" }' ],
+            'invalid name 2' => [ '{ "name": "1foo ", "type": "scalar", "aggregationType": "none" }' ]
+        ];
+    }
+
+    #[\PHPUnit\Framework\Attributes\DataProvider('invalidJson_data')]
+    public function testInvalidJson($input)
+    {
+        $this->expectException(RESTException::class);
+        $this->expectExceptionCode(400);
+        $p = new Product;
+        $se = SchemaEntry::fromJson(json_decode('[ ' . $input . ' ]'), $p);
+    }
+}
+
diff --git a/autotests/schemaentryelementtest.php b/autotests/schemaentryelementtest.php
--- a/autotests/schemaentryelementtest.php
+++ b/autotests/schemaentryelementtest.php
@@ -1,57 +1,55 @@
-<?php
-/*
-    SPDX-FileCopyrightText: 2016 Volker Krause <vkrause@kde.org>
-
-    SPDX-License-Identifier: MIT
-*/
-
-require_once('../src/server/shared/product.php');
-require_once('../src/server/shared/schemaentry.php');
-require_once('../src/server/shared/schemaentryelement.php');
-
-class SchemaEntryElementTest extends PHPUnit\Framework\TestCase
-{
-    public static function dataColumnName_data()
-    {
-        return [
-            'normal' => [ 'foo', 'col_data_entry_foo' ],
-            'dot' => [ 'my.value', 'col_data_entry_my_value' ]
-        ];
-    }
-
-    /** @dataProvider dataColumnName_data */
-    public function testDataTableName($input, $output)
-    {
-        $p = new Product;
-        $p->name = 'org.kde.TestProduct';
-        $se = new SchemaEntry($p);
-        $se->name = 'entry';
-        $see = new SchemaEntryElement($se);
-        $see->name = $input;
-        $this->assertEquals($output, $see->dataColumnName());
-    }
-
-    public static function invalidJson_data()
-    {
-        return [
-            'empty' => [ '{}' ],
-            'empty name' => [ '{ "name": "", "type": "string" }' ],
-            'empty type' => [ '{ "name": "foo", "type": "" }' ],
-            'invalid type' => [ '{ "name": "foo", "type": "bla" }' ],
-            'invalid name' => [ '{ "name": " foo ", "type": "string" }' ],
-            'invalid name 2' => [ '{ "name": "1foo ", "type": "string" }' ]
-        ];
-    }
-
-    /**
-     * @dataProvider invalidJson_data
-     */
-    public function testInvalidJson($input)
-    {
-        $this->expectException(RESTException::class);
-        $this->expectExceptionCode(400);
-        $p = new Product;
-        $se = new SchemaEntry($p);
-        SchemaEntryElement::fromJson(json_decode('[ ' . $input . ' ]'), $se);
-    }
-}
+<?php
+/*
+    SPDX-FileCopyrightText: 2016 Volker Krause <vkrause@kde.org>
+
+    SPDX-License-Identifier: MIT
+*/
+
+require_once('../src/server/shared/product.php');
+require_once('../src/server/shared/schemaentry.php');
+require_once('../src/server/shared/schemaentryelement.php');
+
+class SchemaEntryElementTest extends PHPUnit\Framework\TestCase
+{
+    public static function dataColumnName_data()
+    {
+        return [
+            'normal' => [ 'foo', 'col_data_entry_foo' ],
+            'dot' => [ 'my.value', 'col_data_entry_my_value' ]
+        ];
+    }
+
+    #[\PHPUnit\Framework\Attributes\DataProvider('dataColumnName_data')]
+    public function testDataTableName($input, $output)
+    {
+        $p = new Product;
+        $p->name = 'org.kde.TestProduct';
+        $se = new SchemaEntry($p);
+        $se->name = 'entry';
+        $see = new SchemaEntryElement($se);
+        $see->name = $input;
+        $this->assertEquals($output, $see->dataColumnName());
+    }
+
+    public static function invalidJson_data()
+    {
+        return [
+            'empty' => [ '{}' ],
+            'empty name' => [ '{ "name": "", "type": "string" }' ],
+            'empty type' => [ '{ "name": "foo", "type": "" }' ],
+            'invalid type' => [ '{ "name": "foo", "type": "bla" }' ],
+            'invalid name' => [ '{ "name": " foo ", "type": "string" }' ],
+            'invalid name 2' => [ '{ "name": "1foo ", "type": "string" }' ]
+        ];
+    }
+
+    #[\PHPUnit\Framework\Attributes\DataProvider('invalidJson_data')]
+    public function testInvalidJson($input)
+    {
+        $this->expectException(RESTException::class);
+        $this->expectExceptionCode(400);
+        $p = new Product;
+        $se = new SchemaEntry($p);
+        SchemaEntryElement::fromJson(json_decode('[ ' . $input . ' ]'), $se);
+    }
+}
diff --git a/autotests/producttest.php b/autotests/producttest.php
--- a/autotests/producttest.php
+++ b/autotests/producttest.php
@@ -1,355 +1,353 @@
-<?php
-/*
-    SPDX-FileCopyrightText: 2016 Volker Krause <vkrause@kde.org>
-
-    SPDX-License-Identifier: MIT
-*/
-
-require_once('datastoretesthelper.php');
-require_once('../src/server/shared/product.php');
-
-class ProductTest extends PHPUnit\Framework\TestCase
-{
-    private static $db;
-
-    public static function setUpBeforeClass(): void
-    {
-        self::$db = DatastoreTestHelper::setup();
-    }
-
-    public static function tableName_data()
-    {
-        return [
-            'normal' => [ 'foo', 'pd_foo' ],
-            'dot' => [ 'org.kde.foo', 'pd_org_kde_foo' ]
-        ];
-    }
-
-    /** @dataProvider tableName_data */
-    public function testTableName($input, $output)
-    {
-        $p = new Product;
-        $p->name = $input;
-        $this->assertEquals($output, $p->dataTableName());
-    }
-
-    public function testFromJson()
-    {
-        $p = Product::fromJson('{
-            "name": "org.kde.product",
-            "schema": [
-                {
-                    "name": "entry1",
-                    "type": "scalar",
-                    "elements" : [
-                        { "name": "element1", "type": "int" },
-                        { "name": "element2", "type": "bool" }
-                    ]
-                }
-            ],
-            "aggregation": [ { "type": "category", "name": "n1", "elements": [
-                { "type": "value", "schemaEntry": "entry1", "schemaEntryElement": "element1" }
-            ] } ]
-        }');
-        $this->assertEquals($p->name, 'org.kde.product');
-        $this->assertEquals(1, count($p->schema));
-        $this->assertEquals('pd_org_kde_product', $p->dataTableName());
-        $entry = $p->schema[0];
-        $this->assertEquals($entry->name, 'entry1');
-        $this->assertEquals($entry->type, 'scalar');
-
-        $this->assertEquals(2, count($entry->elements));
-        $elem = $entry->elements[1];
-        $this->assertInstanceOf(SchemaEntryElement::class, $elem);
-        $this->assertEquals('element2', $elem->name);
-        $this->assertEquals('bool', $elem->type);
-
-        $this->assertCount(1, $p->aggregation);
-    }
-
-    public function testToJson()
-    {
-        $p = new Product();
-        $p->name = 'org.kde.product';
-        $entry = new SchemaEntry($p);
-        $entry->name = 'entry1';
-        $entry->type = 'list';
-        $elem = new SchemaEntryElement($entry);
-        $elem->name = 'element3';
-        $elem->type = 'number';
-        array_push($entry->elements, $elem);
-        array_push($p->schema, $entry);
-        $a = new Aggregation;
-        $a->type = 'numeric';
-        $a->name = 'n1';
-        $a->elements = json_decode('[{ "type": "size", "schemaEntry": "entry1" }]');
-        array_push($p->aggregation, $a);
-
-        $this->assertJsonStringEqualsJsonString(json_encode($p), '{
-            "name": "org.kde.product",
-            "schema": [
-                {
-                    "name": "entry1",
-                    "type": "list",
-                    "elements" : [
-                        { "name": "element3", "type": "number" }
-                    ]
-                }
-            ],
-            "aggregation": [ { "type": "numeric", "name": "n1", "elements": [
-                { "type": "size", "schemaEntry": "entry1" }
-            ] } ]
-        }');
-    }
-
-    public function testAllProducts()
-    {
-        $ps = Product::allProducts(self::$db);
-        $this->assertEquals(1, count($ps));
-        $p = $ps[0];
-        $this->assertInstanceOf(Product::class, $p);
-        $this->assertEquals('org.kde.UnitTest', $p->name);
-        $this->assertEquals(2, count($p->schema));
-
-        $entry1 = $p->schema[0];
-        $entry2 = $p->schema[1];
-        if ($entry1->name > $entry2->name) {
-            $entry1 = $p->schema[1];
-            $entry2 = $p->schema[0];
-        }
-
-        $this->assertInstanceOf(SchemaEntry::class, $entry1);
-        $this->assertEquals($entry1->name, 'entry1');
-        $this->assertEquals($entry1->type, 'scalar');
-        $this->assertEquals(2, count($entry1->elements));
-        $elem = $entry1->elements[1];
-        $this->assertInstanceOf(SchemaEntryElement::class, $elem);
-        $this->assertEquals('element12', $elem->name);
-        $this->assertEquals('bool', $elem->type);
-
-        $this->assertInstanceOf(SchemaEntry::class, $entry2);
-        $this->assertEquals($entry2->name, 'entry2');
-        $this->assertEquals($entry2->type, 'list');
-
-        $this->assertCount(2, $p->aggregation);
-    }
-
-    public function testProductByName()
-    {
-        $p = Product::productByName(self::$db, 'I don\'t exist');
-        $this->assertNull($p);
-
-        $p = Product::productByName(self::$db, 'org.kde.UnitTest');
-        $this->assertInstanceOf(Product::class, $p);
-        $this->assertEquals('org.kde.UnitTest', $p->name);
-        $this->assertEquals(2, count($p->schema));
-
-        $entry1 = $p->schema[0];
-        $entry2 = $p->schema[1];
-        if ($entry1->name > $entry2->name) {
-            $entry1 = $p->schema[1];
-            $entry2 = $p->schema[0];
-        }
-
-        $this->assertInstanceOf(SchemaEntry::class, $entry1);
-        $this->assertEquals($entry1->name, 'entry1');
-        $this->assertEquals($entry1->type, 'scalar');
-        $elem = $entry1->elements[1];
-        $this->assertInstanceOf(SchemaEntryElement::class, $elem);
-        $this->assertEquals('element12', $elem->name);
-        $this->assertEquals('bool', $elem->type);
-
-        $this->assertInstanceOf(SchemaEntry::class, $entry2);
-        $this->assertEquals($entry2->name, 'entry2');
-        $this->assertEquals($entry2->type, 'list');
-    }
-
-    public function testProductInsert()
-    {
-        $json = '{
-            "name": "org.kde.NewProduct",
-            "schema": [
-                {
-                    "name": "entryA",
-                    "type": "scalar",
-                    "elements" : [
-                        { "name": "elementA1", "type": "int" },
-                        { "name": "elementA2", "type": "bool" }
-                    ]
-                },
-                {
-                    "name": "entryB",
-                    "type": "map",
-                    "elements" : [
-                        { "name": "elementB1", "type": "number" },
-                        { "name": "elementB2", "type": "string" }
-                    ]
-                }
-            ],
-            "aggregation": [
-                {
-                    "type": "category",
-                    "name": "n1",
-                    "elements": [
-                        { "type": "value", "schemaEntry": "entryA", "schemaEntryElement": "elementA1" }
-                    ]
-                },
-                {
-                    "type": "ratio_set",
-                    "name": "n2",
-                    "elements": [
-                        { "type": "value", "schemaEntry": "entryB", "schemaEntryElement": "elementB1" }
-                    ]
-                }
-            ]
-        }';
-
-        $p = Product::fromJson($json);
-        $p->insert(self::$db);
-
-        $p = Product::productByName(self::$db, 'org.kde.NewProduct');
-        $p->aggregation = Aggregation::aggregationsForProduct(self::$db, $p);
-        $this->assertJsonStringEqualsJsonString($json, json_encode($p));
-    }
-
-    public function testProductUpdate()
-    {
-        $jsonOld = '{
-            "name": "org.kde.ProductToUpdate",
-            "schema": [
-                {
-                    "name": "entryA",
-                    "type": "scalar",
-                    "elements" : [
-                        { "name": "elementA1", "type": "int" },
-                        { "name": "elementA2", "type": "bool" }
-                    ]
-                },
-                {
-                    "name": "entryB",
-                    "type": "list",
-                    "elements" : [
-                        { "name": "elementB1", "type": "number" },
-                        { "name": "elementB2", "type": "string" }
-                    ]
-                }
-            ],
-            "aggregation": [
-                {
-                    "type": "ratio_set",
-                    "name": "n2",
-                    "elements": [
-                        { "type": "value", "schemaEntry": "entryB", "schemaEntryElement": "elementB1" }
-                    ]
-                }
-            ]
-        }';
-
-        $p = Product::fromJson($jsonOld);
-        $p->insert(self::$db);
-
-        $p = Product::productByName(self::$db, 'org.kde.ProductToUpdate');
-        $p->aggregation = Aggregation::aggregationsForProduct(self::$db, $p);
-        $this->assertJsonStringEqualsJsonString($jsonOld, json_encode($p));
-
-        $jsonNew = '{
-            "name": "org.kde.ProductToUpdate",
-            "schema": [
-                {
-                    "name": "entryA",
-                    "type": "scalar",
-                    "elements" : [
-                        { "name": "elementA1", "type": "int" }
-                    ]
-                },
-                {
-                    "name": "entryC",
-                    "type": "map",
-                    "elements" : [
-                        { "name": "elementC1", "type": "number" },
-                        { "name": "elementC2", "type": "number" }
-                    ]
-                }
-            ],
-            "aggregation": [
-                {
-                    "type": "category",
-                    "name": "n1",
-                    "elements": [
-                        { "type": "value", "schemaEntry": "entryA", "schemaEntryElement": "elementA1" }
-                    ]
-                },
-                {
-                    "type": "xy",
-                    "name": "n3",
-                    "elements": [
-                        { "type": "value", "schemaEntry": "entryC", "schemaEntryElement": "elementC1" },
-                        { "type": "value", "schemaEntry": "entryC", "schemaEntryElement": "elementC2" }
-                    ]
-                }
-            ]
-        }';
-        $newP = Product::fromJson($jsonNew);
-        $p->update(self::$db, $newP);
-
-        $p = Product::productByName(self::$db, 'org.kde.ProductToUpdate');
-        $p->aggregation = Aggregation::aggregationsForProduct(self::$db, $p);
-        $this->assertJsonStringEqualsJsonString($jsonNew, json_encode($p));
-    }
-
-    public function testProductDelete()
-    {
-        $json = '{
-            "name": "org.kde.DeletedProduct",
-            "schema": [
-                {
-                    "name": "entry1",
-                    "type": "list",
-                    "elements" : [
-                        { "name": "element3", "type": "number" }
-                    ]
-                }
-            ],
-            "aggregation": [
-                {
-                    "type": "numeric",
-                    "name": "n4",
-                    "elements": [
-                        { "type": "size", "schemaEntry": "entry1" }
-                    ]
-                }
-            ]
-        }';
-
-        $p = Product::fromJson($json);
-        $p->insert(self::$db);
-        $p = Product::productByName(self::$db, 'org.kde.DeletedProduct');
-        $p->aggregation = Aggregation::aggregationsForProduct(self::$db, $p);
-        $this->assertJsonStringEqualsJsonString($json, json_encode($p));
-
-        $p->delete(self::$db);
-        $p = Product::productByName(self::$db, 'org.kde.DeletedProduct');
-        $this->assertNull($p);
-    }
-
-    public static function invalidInput_data()
-    {
-        return [
-            'empty' => [ '{}' ],
-            'empty name' => [ '{ "name": "", "schema": [] }' ],
-            'invalid name' => [ '{ "name": "123", "schema": [] }' ],
-            'non-string name' => [ '{ "name": 123, "schema": [] }' ]
-        ];
-    }
-
-    /**
-     * @dataProvider invalidInput_data
-     */
-    public function testInvalidInput($json)
-    {
-        $this->expectException(RESTException::class);
-        $this->expectExceptionCode(400);
-        $p = Product::fromJson($json);
-    }
-}
-
+<?php
+/*
+    SPDX-FileCopyrightText: 2016 Volker Krause <vkrause@kde.org>
+
+    SPDX-License-Identifier: MIT
+*/
+
+require_once('datastoretesthelper.php');
+require_once('../src/server/shared/product.php');
+
+class ProductTest extends PHPUnit\Framework\TestCase
+{
+    private static $db;
+
+    public static function setUpBeforeClass(): void
+    {
+        self::$db = DatastoreTestHelper::setup();
+    }
+
+    public static function tableName_data()
+    {
+        return [
+            'normal' => [ 'foo', 'pd_foo' ],
+            'dot' => [ 'org.kde.foo', 'pd_org_kde_foo' ]
+        ];
+    }
+
+    #[\PHPUnit\Framework\Attributes\DataProvider('tableName_data')]
+    public function testTableName($input, $output)
+    {
+        $p = new Product;
+        $p->name = $input;
+        $this->assertEquals($output, $p->dataTableName());
+    }
+
+    public function testFromJson()
+    {
+        $p = Product::fromJson('{
+            "name": "org.kde.product",
+            "schema": [
+                {
+                    "name": "entry1",
+                    "type": "scalar",
+                    "elements" : [
+                        { "name": "element1", "type": "int" },
+                        { "name": "element2", "type": "bool" }
+                    ]
+                }
+            ],
+            "aggregation": [ { "type": "category", "name": "n1", "elements": [
+                { "type": "value", "schemaEntry": "entry1", "schemaEntryElement": "element1" }
+            ] } ]
+        }');
+        $this->assertEquals($p->name, 'org.kde.product');
+        $this->assertEquals(1, count($p->schema));
+        $this->assertEquals('pd_org_kde_product', $p->dataTableName());
+        $entry = $p->schema[0];
+        $this->assertEquals($entry->name, 'entry1');
+        $this->assertEquals($entry->type, 'scalar');
+
+        $this->assertEquals(2, count($entry->elements));
+        $elem = $entry->elements[1];
+        $this->assertInstanceOf(SchemaEntryElement::class, $elem);
+        $this->assertEquals('element2', $elem->name);
+        $this->assertEquals('bool', $elem->type);
+
+        $this->assertCount(1, $p->aggregation);
+    }
+
+    public function testToJson()
+    {
+        $p = new Product();
+        $p->name = 'org.kde.product';
+        $entry = new SchemaEntry($p);
+        $entry->name = 'entry1';
+        $entry->type = 'list';
+        $elem = new SchemaEntryElement($entry);
+        $elem->name = 'element3';
+        $elem->type = 'number';
+        array_push($entry->elements, $elem);
+        array_push($p->schema, $entry);
+        $a = new Aggregation;
+        $a->type = 'numeric';
+        $a->name = 'n1';
+        $a->elements = json_decode('[{ "type": "size", "schemaEntry": "entry1" }]');
+        array_push($p->aggregation, $a);
+
+        $this->assertJsonStringEqualsJsonString(json_encode($p), '{
+            "name": "org.kde.product",
+            "schema": [
+                {
+                    "name": "entry1",
+                    "type": "list",
+                    "elements" : [
+                        { "name": "element3", "type": "number" }
+                    ]
+                }
+            ],
+            "aggregation": [ { "type": "numeric", "name": "n1", "elements": [
+                { "type": "size", "schemaEntry": "entry1" }
+            ] } ]
+        }');
+    }
+
+    public function testAllProducts()
+    {
+        $ps = Product::allProducts(self::$db);
+        $this->assertEquals(1, count($ps));
+        $p = $ps[0];
+        $this->assertInstanceOf(Product::class, $p);
+        $this->assertEquals('org.kde.UnitTest', $p->name);
+        $this->assertEquals(2, count($p->schema));
+
+        $entry1 = $p->schema[0];
+        $entry2 = $p->schema[1];
+        if ($entry1->name > $entry2->name) {
+            $entry1 = $p->schema[1];
+            $entry2 = $p->schema[0];
+        }
+
+        $this->assertInstanceOf(SchemaEntry::class, $entry1);
+        $this->assertEquals($entry1->name, 'entry1');
+        $this->assertEquals($entry1->type, 'scalar');
+        $this->assertEquals(2, count($entry1->elements));
+        $elem = $entry1->elements[1];
+        $this->assertInstanceOf(SchemaEntryElement::class, $elem);
+        $this->assertEquals('element12', $elem->name);
+        $this->assertEquals('bool', $elem->type);
+
+        $this->assertInstanceOf(SchemaEntry::class, $entry2);
+        $this->assertEquals($entry2->name, 'entry2');
+        $this->assertEquals($entry2->type, 'list');
+
+        $this->assertCount(2, $p->aggregation);
+    }
+
+    public function testProductByName()
+    {
+        $p = Product::productByName(self::$db, 'I don\'t exist');
+        $this->assertNull($p);
+
+        $p = Product::productByName(self::$db, 'org.kde.UnitTest');
+        $this->assertInstanceOf(Product::class, $p);
+        $this->assertEquals('org.kde.UnitTest', $p->name);
+        $this->assertEquals(2, count($p->schema));
+
+        $entry1 = $p->schema[0];
+        $entry2 = $p->schema[1];
+        if ($entry1->name > $entry2->name) {
+            $entry1 = $p->schema[1];
+            $entry2 = $p->schema[0];
+        }
+
+        $this->assertInstanceOf(SchemaEntry::class, $entry1);
+        $this->assertEquals($entry1->name, 'entry1');
+        $this->assertEquals($entry1->type, 'scalar');
+        $elem = $entry1->elements[1];
+        $this->assertInstanceOf(SchemaEntryElement::class, $elem);
+        $this->assertEquals('element12', $elem->name);
+        $this->assertEquals('bool', $elem->type);
+
+        $this->assertInstanceOf(SchemaEntry::class, $entry2);
+        $this->assertEquals($entry2->name, 'entry2');
+        $this->assertEquals($entry2->type, 'list');
+    }
+
+    public function testProductInsert()
+    {
+        $json = '{
+            "name": "org.kde.NewProduct",
+            "schema": [
+                {
+                    "name": "entryA",
+                    "type": "scalar",
+                    "elements" : [
+                        { "name": "elementA1", "type": "int" },
+                        { "name": "elementA2", "type": "bool" }
+                    ]
+                },
+                {
+                    "name": "entryB",
+                    "type": "map",
+                    "elements" : [
+                        { "name": "elementB1", "type": "number" },
+                        { "name": "elementB2", "type": "string" }
+                    ]
+                }
+            ],
+            "aggregation": [
+                {
+                    "type": "category",
+                    "name": "n1",
+                    "elements": [
+                        { "type": "value", "schemaEntry": "entryA", "schemaEntryElement": "elementA1" }
+                    ]
+                },
+                {
+                    "type": "ratio_set",
+                    "name": "n2",
+                    "elements": [
+                        { "type": "value", "schemaEntry": "entryB", "schemaEntryElement": "elementB1" }
+                    ]
+                }
+            ]
+        }';
+
+        $p = Product::fromJson($json);
+        $p->insert(self::$db);
+
+        $p = Product::productByName(self::$db, 'org.kde.NewProduct');
+        $p->aggregation = Aggregation::aggregationsForProduct(self::$db, $p);
+        $this->assertJsonStringEqualsJsonString($json, json_encode($p));
+    }
+
+    public function testProductUpdate()
+    {
+        $jsonOld = '{
+            "name": "org.kde.ProductToUpdate",
+            "schema": [
+                {
+                    "name": "entryA",
+                    "type": "scalar",
+                    "elements" : [
+                        { "name": "elementA1", "type": "int" },
+                        { "name": "elementA2", "type": "bool" }
+                    ]
+                },
+                {
+                    "name": "entryB",
+                    "type": "list",
+                    "elements" : [
+                        { "name": "elementB1", "type": "number" },
+                        { "name": "elementB2", "type": "string" }
+                    ]
+                }
+            ],
+            "aggregation": [
+                {
+                    "type": "ratio_set",
+                    "name": "n2",
+                    "elements": [
+                        { "type": "value", "schemaEntry": "entryB", "schemaEntryElement": "elementB1" }
+                    ]
+                }
+            ]
+        }';
+
+        $p = Product::fromJson($jsonOld);
+        $p->insert(self::$db);
+
+        $p = Product::productByName(self::$db, 'org.kde.ProductToUpdate');
+        $p->aggregation = Aggregation::aggregationsForProduct(self::$db, $p);
+        $this->assertJsonStringEqualsJsonString($jsonOld, json_encode($p));
+
+        $jsonNew = '{
+            "name": "org.kde.ProductToUpdate",
+            "schema": [
+                {
+                    "name": "entryA",
+                    "type": "scalar",
+                    "elements" : [
+                        { "name": "elementA1", "type": "int" }
+                    ]
+                },
+                {
+                    "name": "entryC",
+                    "type": "map",
+                    "elements" : [
+                        { "name": "elementC1", "type": "number" },
+                        { "name": "elementC2", "type": "number" }
+                    ]
+                }
+            ],
+            "aggregation": [
+                {
+                    "type": "category",
+                    "name": "n1",
+                    "elements": [
+                        { "type": "value", "schemaEntry": "entryA", "schemaEntryElement": "elementA1" }
+                    ]
+                },
+                {
+                    "type": "xy",
+                    "name": "n3",
+                    "elements": [
+                        { "type": "value", "schemaEntry": "entryC", "schemaEntryElement": "elementC1" },
+                        { "type": "value", "schemaEntry": "entryC", "schemaEntryElement": "elementC2" }
+                    ]
+                }
+            ]
+        }';
+        $newP = Product::fromJson($jsonNew);
+        $p->update(self::$db, $newP);
+
+        $p = Product::productByName(self::$db, 'org.kde.ProductToUpdate');
+        $p->aggregation = Aggregation::aggregationsForProduct(self::$db, $p);
+        $this->assertJsonStringEqualsJsonString($jsonNew, json_encode($p));
+    }
+
+    public function testProductDelete()
+    {
+        $json = '{
+            "name": "org.kde.DeletedProduct",
+            "schema": [
+                {
+                    "name": "entry1",
+                    "type": "list",
+                    "elements" : [
+                        { "name": "element3", "type": "number" }
+                    ]
+                }
+            ],
+            "aggregation": [
+                {
+                    "type": "numeric",
+                    "name": "n4",
+                    "elements": [
+                        { "type": "size", "schemaEntry": "entry1" }
+                    ]
+                }
+            ]
+        }';
+
+        $p = Product::fromJson($json);
+        $p->insert(self::$db);
+        $p = Product::productByName(self::$db, 'org.kde.DeletedProduct');
+        $p->aggregation = Aggregation::aggregationsForProduct(self::$db, $p);
+        $this->assertJsonStringEqualsJsonString($json, json_encode($p));
+
+        $p->delete(self::$db);
+        $p = Product::productByName(self::$db, 'org.kde.DeletedProduct');
+        $this->assertNull($p);
+    }
+
+    public static function invalidInput_data()
+    {
+        return [
+            'empty' => [ '{}' ],
+            'empty name' => [ '{ "name": "", "schema": [] }' ],
+            'invalid name' => [ '{ "name": "123", "schema": [] }' ],
+            'non-string name' => [ '{ "name": 123, "schema": [] }' ]
+        ];
+    }
+
+    #[\PHPUnit\Framework\Attributes\DataProvider('invalidInput_data')]
+    public function testInvalidInput($json)
+    {
+        $this->expectException(RESTException::class);
+        $this->expectExceptionCode(400);
+        $p = Product::fromJson($json);
+    }
+}
+
diff --git a/autotests/sampletest.php b/autotests/sampletest.php
--- a/autotests/sampletest.php
+++ b/autotests/sampletest.php
@@ -1,183 +1,179 @@
-<?php
-/*
-    SPDX-FileCopyrightText: 2016 Volker Krause <vkrause@kde.org>
-
-    SPDX-License-Identifier: MIT
-*/
-
-require_once('datastoretesthelper.php');
-require_once('../src/server/shared/product.php');
-require_once('../src/server/shared/sample.php');
-
-class SampleTest extends PHPUnit\Framework\TestCase
-{
-    private static $db;
-
-    public static function setUpBeforeClass(): void
-    {
-        self::$db = DatastoreTestHelper::setup();
-    }
-
-    public function testListSampleInsert()
-    {
-        $p = Product::productByName(self::$db, 'org.kde.UnitTest');
-        $this->assertNotNull($p);
-        $p->name = 'org.kde.MyListProduct';
-        $p->insert(self::$db); // HACK create a new product, so the data tables are created correctly
-
-        $sample = '{
-            "entry1": {
-                "element11": "aString",
-                "element12": true
-            },
-            "entry2": [
-                { "element21": 14, "element22": 1.5 },
-                { "element21": 16, "element22": 1.7 }
-            ]
-        }';
-
-        Sample::insert(self::$db, $sample, $p);
-        Sample::insert(self::$db, $sample, $p);
-
-        Sample::echoDataAsJson(self::$db, $p);
-        $data = json_decode($this->getActualOutput(), false);
-        $this->assertTrue(is_array($data));
-        $this->assertCount(2, $data);
-        $d0 = $data[0];
-        $this->assertObjectHasAttribute('timestamp', $d0);
-        $this->assertObjectHasAttribute('entry1', $d0);
-        $this->assertObjectHasAttribute('entry2', $d0);
-        $d01 = $d0->{'entry2'};
-        $this->assertCount(2, $d01);
-    }
-
-    public function testMapSampleInsert()
-    {
-        $p = Product::productByName(self::$db, 'org.kde.UnitTest');
-        $this->assertNotNull($p);
-        $p->name = 'org.kde.MyMapProduct';
-        $p->schema[1]->type = SchemaEntry::MAP_TYPE;
-        $p->insert(self::$db);
-
-        $sample = '{
-            "entry1": {
-                "element11": "aString",
-                "element12": true
-            },
-            "entry2": {
-                "key1": { "element21": 14, "element22": 1.5 },
-                "key2": { "element21": 16, "element22": 1.7 }
-            }
-        }';
-
-        Sample::insert(self::$db, $sample, $p);
-
-        Sample::echoDataAsJson(self::$db, $p);
-        $data = json_decode($this->getActualOutput(), false);
-
-        $this->assertTrue(is_array($data));
-        $this->assertCount(1, $data);
-        $d0 = $data[0];
-        $this->assertObjectHasAttribute('timestamp', $d0);
-        $this->assertObjectHasAttribute('entry1', $d0);
-        $this->assertObjectHasAttribute('entry2', $d0);
-        $d01 = $d0->{'entry2'};
-        $this->assertObjectHasAttribute('key1', $d01);
-        $this->assertObjectHasAttribute('key2', $d01);
-    }
-
-    public function testEmptyInsert()
-    {
-        $p = Product::productByName(self::$db, 'org.kde.UnitTest');
-        $this->assertNotNull($p);
-        $p->name = 'org.kde.MyEmptyProduct';
-        $p->insert(self::$db);
-
-        $sample = '{
-            "someRandomStuff": "not part of the schema",
-            "someOtherStuff": 42
-        }';
-
-        Sample::insert(self::$db, $sample, $p);
-        Sample::echoDataAsJson(self::$db, $p);
-        $data = json_decode($this->getActualOutput(), false);
-        $this->assertTrue(is_array($data));
-        $this->assertCount(1, $data);
-        $d0 = $data[0];
-        $this->assertObjectHasAttribute('timestamp', $d0);
-        $this->assertObjectNotHasAttribute('entry1', $d0);
-        $this->assertObjectNotHasAttribute('someRandomStuff', $d0);
-    }
-
-    public static function invalidInsert_data()
-    {
-        return [
-            'empty' => [ '' ],
-            'array' => [ '[]' ],
-            'missing id' => [ '{ "timestamp": "2016-12-18 12:42:35" }' ],
-            'missing timestamp' => [ '{ "id": 42 }' ]
-        ];
-    }
-
-    /**
-     * @dataProvider invalidInsert_data
-     */
-    public function testInvalidInsert($input)
-    {
-        $this->expectException(RESTException::class);
-        $this->expectExceptionCode(400);
-        $p = Product::productByName(self::$db, 'org.kde.UnitTest');
-        $this->assertNotNull($p);
-
-        Sample::insert(self::$db, $input, $p);
-    }
-
-    public function testImport()
-    {
-        $p = Product::productByName(self::$db, 'org.kde.UnitTest');
-        $this->assertNotNull($p);
-        $p->name = 'org.kde.MyCleanProduct';
-        $p->insert(self::$db);
-
-        $input = '[{
-            "timestamp": "2016-12-18 12:42:35",
-            "entry1": { "element11": "firstString" },
-            "entry2": [ { "element21": 12 } ]
-        }, {
-            "timestamp": "2016-12-19 15:12:10",
-            "entry1": { "element12": true },
-            "entry2": [ { "element22": 1.3 } ]
-        }]';
-        Sample::import(self::$db, $input, $p);
-
-        Sample::echoDataAsJson(self::$db, $p);
-        $jsonOutput = json_decode($this->getActualOutput(), true);
-        unset($jsonOutput[0]['id']);
-        unset($jsonOutput[1]['id']);
-        $this->assertJsonStringEqualsJsonString($input, json_encode($jsonOutput));
-    }
-
-    public static function invalidImport_data()
-    {
-        return [
-            'nothing' => [ '' ],
-            'empty' => [ '{}' ],
-            'object' => [ '{ "id": 42, "timestamp": "2016-12-18 12:42:35" }' ],
-            'missing timestamp' => [ '[{ "id": 42 }]' ],
-            'id present' => [ '[{ "id": 42, "timestamp": "2016-12-18 12:42:35" }]' ]
-        ];
-    }
-
-    /**
-     * @dataProvider invalidImport_data
-     */
-    public function testInvalidImport($input)
-    {
-        $this->expectException(RESTException::class);
-        $this->expectExceptionCode(400);
-        $p = Product::productByName(self::$db, 'org.kde.UnitTest');
-        $this->assertNotNull($p);
-
-        Sample::import(self::$db, $input, $p);
-    }
-}
+<?php
+/*
+    SPDX-FileCopyrightText: 2016 Volker Krause <vkrause@kde.org>
+
+    SPDX-License-Identifier: MIT
+*/
+
+require_once('datastoretesthelper.php');
+require_once('../src/server/shared/product.php');
+require_once('../src/server/shared/sample.php');
+
+class SampleTest extends PHPUnit\Framework\TestCase
+{
+    private static $db;
+
+    public static function setUpBeforeClass(): void
+    {
+        self::$db = DatastoreTestHelper::setup();
+    }
+
+    public function testListSampleInsert()
+    {
+        $p = Product::productByName(self::$db, 'org.kde.UnitTest');
+        $this->assertNotNull($p);
+        $p->name = 'org.kde.MyListProduct';
+        $p->insert(self::$db); // HACK create a new product, so the data tables are created correctly
+
+        $sample = '{
+            "entry1": {
+                "element11": "aString",
+                "element12": true
+            },
+            "entry2": [
+                { "element21": 14, "element22": 1.5 },
+                { "element21": 16, "element22": 1.7 }
+            ]
+        }';
+
+        Sample::insert(self::$db, $sample, $p);
+        Sample::insert(self::$db, $sample, $p);
+
+        Sample::echoDataAsJson(self::$db, $p);
+        $data = json_decode($this->getActualOutput(), false);
+        $this->assertTrue(is_array($data));
+        $this->assertCount(2, $data);
+        $d0 = $data[0];
+        $this->assertObjectHasAttribute('timestamp', $d0);
+        $this->assertObjectHasAttribute('entry1', $d0);
+        $this->assertObjectHasAttribute('entry2', $d0);
+        $d01 = $d0->{'entry2'};
+        $this->assertCount(2, $d01);
+    }
+
+    public function testMapSampleInsert()
+    {
+        $p = Product::productByName(self::$db, 'org.kde.UnitTest');
+        $this->assertNotNull($p);
+        $p->name = 'org.kde.MyMapProduct';
+        $p->schema[1]->type = SchemaEntry::MAP_TYPE;
+        $p->insert(self::$db);
+
+        $sample = '{
+            "entry1": {
+                "element11": "aString",
+                "element12": true
+            },
+            "entry2": {
+                "key1": { "element21": 14, "element22": 1.5 },
+                "key2": { "element21": 16, "element22": 1.7 }
+            }
+        }';
+
+        Sample::insert(self::$db, $sample, $p);
+
+        Sample::echoDataAsJson(self::$db, $p);
+        $data = json_decode($this->getActualOutput(), false);
+
+        $this->assertTrue(is_array($data));
+        $this->assertCount(1, $data);
+        $d0 = $data[0];
+        $this->assertObjectHasAttribute('timestamp', $d0);
+        $this->assertObjectHasAttribute('entry1', $d0);
+        $this->assertObjectHasAttribute('entry2', $d0);
+        $d01 = $d0->{'entry2'};
+        $this->assertObjectHasAttribute('key1', $d01);
+        $this->assertObjectHasAttribute('key2', $d01);
+    }
+
+    public function testEmptyInsert()
+    {
+        $p = Product::productByName(self::$db, 'org.kde.UnitTest');
+        $this->assertNotNull($p);
+        $p->name = 'org.kde.MyEmptyProduct';
+        $p->insert(self::$db);
+
+        $sample = '{
+            "someRandomStuff": "not part of the schema",
+            "someOtherStuff": 42
+        }';
+
+        Sample::insert(self::$db, $sample, $p);
+        Sample::echoDataAsJson(self::$db, $p);
+        $data = json_decode($this->getActualOutput(), false);
+        $this->assertTrue(is_array($data));
+        $this->assertCount(1, $data);
+        $d0 = $data[0];
+        $this->assertObjectHasAttribute('timestamp', $d0);
+        $this->assertObjectNotHasAttribute('entry1', $d0);
+        $this->assertObjectNotHasAttribute('someRandomStuff', $d0);
+    }
+
+    public static function invalidInsert_data()
+    {
+        return [
+            'empty' => [ '' ],
+            'array' => [ '[]' ],
+            'missing id' => [ '{ "timestamp": "2016-12-18 12:42:35" }' ],
+            'missing timestamp' => [ '{ "id": 42 }' ]
+        ];
+    }
+
+    #[\PHPUnit\Framework\Attributes\DataProvider('invalidInsert_data')]
+    public function testInvalidInsert($input)
+    {
+        $this->expectException(RESTException::class);
+        $this->expectExceptionCode(400);
+        $p = Product::productByName(self::$db, 'org.kde.UnitTest');
+        $this->assertNotNull($p);
+
+        Sample::insert(self::$db, $input, $p);
+    }
+
+    public function testImport()
+    {
+        $p = Product::productByName(self::$db, 'org.kde.UnitTest');
+        $this->assertNotNull($p);
+        $p->name = 'org.kde.MyCleanProduct';
+        $p->insert(self::$db);
+
+        $input = '[{
+            "timestamp": "2016-12-18 12:42:35",
+            "entry1": { "element11": "firstString" },
+            "entry2": [ { "element21": 12 } ]
+        }, {
+            "timestamp": "2016-12-19 15:12:10",
+            "entry1": { "element12": true },
+            "entry2": [ { "element22": 1.3 } ]
+        }]';
+        Sample::import(self::$db, $input, $p);
+
+        Sample::echoDataAsJson(self::$db, $p);
+        $jsonOutput = json_decode($this->getActualOutput(), true);
+        unset($jsonOutput[0]['id']);
+        unset($jsonOutput[1]['id']);
+        $this->assertJsonStringEqualsJsonString($input, json_encode($jsonOutput));
+    }
+
+    public static function invalidImport_data()
+    {
+        return [
+            'nothing' => [ '' ],
+            'empty' => [ '{}' ],
+            'object' => [ '{ "id": 42, "timestamp": "2016-12-18 12:42:35" }' ],
+            'missing timestamp' => [ '[{ "id": 42 }]' ],
+            'id present' => [ '[{ "id": 42, "timestamp": "2016-12-18 12:42:35" }]' ]
+        ];
+    }
+
+    #[\PHPUnit\Framework\Attributes\DataProvider('invalidImport_data')]
+    public function testInvalidImport($input)
+    {
+        $this->expectException(RESTException::class);
+        $this->expectExceptionCode(400);
+        $p = Product::productByName(self::$db, 'org.kde.UnitTest');
+        $this->assertNotNull($p);
+
+        Sample::import(self::$db, $input, $p);
+    }
+}
diff --git a/autotests/surveytest.php b/autotests/surveytest.php
--- a/autotests/surveytest.php
+++ b/autotests/surveytest.php
@@ -1,202 +1,200 @@
-<?php
-/*
-    SPDX-FileCopyrightText: 2016 Volker Krause <vkrause@kde.org>
-
-    SPDX-License-Identifier: MIT
-*/
-
-require_once('datastoretesthelper.php');
-require_once('../src/server/shared/product.php');
-require_once('../src/server/shared/survey.php');
-
-class SurveyTest extends PHPUnit\Framework\TestCase
-{
-    private static $db;
-
-    public static function setUpBeforeClass(): void
-    {
-        self::$db = DatastoreTestHelper::setup();
-    }
-
-    public function testFromJson()
-    {
-        $s = Survey::fromJson('{
-            "name": "mySurvey",
-            "url": "http://survey.example/",
-            "active": true,
-            "target": "screen[0].dpi > 100",
-            "uuid": "{9e529dfa-0213-413e-a1a8-8a9cea7d5a97}"
-        }');
-        $this->assertEquals($s->name, 'mySurvey');
-        $this->assertEquals($s->url, 'http://survey.example/');
-        $this->assertEquals($s->active, true);
-        $this->assertEquals($s->uuid, '{9e529dfa-0213-413e-a1a8-8a9cea7d5a97}');
-        $this->assertEquals($s->target, 'screen[0].dpi > 100');
-    }
-
-    public function testToJson()
-    {
-        $s = new Survey();
-        $s->name = 'mySurvey';
-        $s->url = 'http://survey.example/';
-        $s->active = true;
-        $s->uuid = '{9e529dfa-0213-413e-a1a8-8a9cea7d5a97}';
-        $s->target = 'screen[0].dpi > 100';
-
-        $this->assertJsonStringEqualsJsonString(json_encode($s), '{
-            "uuid": "{9e529dfa-0213-413e-a1a8-8a9cea7d5a97}",
-            "name": "mySurvey",
-            "url": "http://survey.example/",
-            "active": true,
-            "target": "screen[0].dpi > 100"
-        }');
-    }
-
-    public function testGetSurveys()
-    {
-        $surveys = Survey::surveysForProduct(self::$db, 'I don\'t exist');
-        $this->assertCount(0, $surveys);
-
-        $surveys = Survey::surveysForProduct(self::$db, 'org.kde.UnitTest');
-        $this->assertCount(2, $surveys);
-
-        $survey1 = $surveys[0];
-        $this->assertInstanceOf(Survey::class, $survey1);
-        $survey2 = $surveys[1];
-        $this->assertInstanceOf(Survey::class, $survey2);
-
-        if ($survey1->name > $survey2->name) {
-            $survey1 = $surveys->schema[1];
-            $survey2 = $surveys->schema[0];
-        }
-
-        $this->assertEquals($survey1->name, 'myActiveSurvey');
-        $this->assertEquals($survey1->url, 'http://survey.example/active');
-        $this->assertEquals($survey1->active, true);
-        $this->assertEquals($survey1->target, 'usageTime.value >= 3600');
-
-        $this->assertEquals($survey2->name, 'myInactiveSurvey');
-        $this->assertEquals($survey2->url, 'http://survey.example/inactive');
-        $this->assertEquals($survey2->active, false);
-        $this->assertEquals($survey2->target, 'screen[0].dpi < 200');
-
-        $p = Product::productByName(self::$db, 'org.kde.UnitTest');
-        $this->assertNotNull($p);
-        $surveys = Survey::activeSurveysForProduct(self::$db, $p);
-        $this->assertCount(1, $surveys);
-
-        $s = $surveys[0];
-        $this->assertInstanceOf(Survey::class, $s);
-        $this->assertEquals($s->name, 'myActiveSurvey');
-        $this->assertEquals($s->url, 'http://survey.example/active');
-        $this->assertEquals($s->active, true);
-        $this->assertEquals($s->target, 'usageTime.value >= 3600');
-    }
-
-    public function testSurveyInsert()
-    {
-        $json = '{
-            "name": "newSurvey",
-            "uuid": "{9e529dfa-0213-413e-a1a8-8a9cea7d5a97}",
-            "url": "http://survey.example/new",
-            "active": true,
-            "target": "startCount.value > 10"
-        }';
-
-        $p = Product::productByName(self::$db, 'org.kde.UnitTest');
-        $this->assertNotNull($p);
-
-        $s = Survey::fromJson($json);
-        $s->insert(self::$db, $p);
-
-        $surveys = Survey::activeSurveysForProduct(self::$db, $p);
-        $s = null;
-        foreach ($surveys as $survey) {
-            if ($survey->name == 'newSurvey') {
-                $s = $survey;
-                break;
-            }
-        }
-        $this->assertNotNull($s);
-        $this->assertEquals($s->name, 'newSurvey');
-        $this->assertEquals($s->url, 'http://survey.example/new');
-        $this->assertEquals($s->active, true);
-        $this->assertEquals($s->target, 'startCount.value > 10');
-        $this->assertEquals($s->uuid, '{9e529dfa-0213-413e-a1a8-8a9cea7d5a97}');
-    }
-
-    public function testSurveyUpdate()
-    {
-        $json = '{
-            "name": "myChangedSurvey",
-            "url": "http://survey.example/changed",
-            "active": false,
-            "uuid": "{962bbd80-7120-4f18-a4c0-5800fa323868}",
-            "target": "views[\"editor\"].ratio > 0.5"
-        }';
-
-        $s = Survey::fromJson($json);
-        $s->update(self::$db);
-
-        $surveys = Survey::surveysForProduct(self::$db, 'org.kde.UnitTest');
-        $s = null;
-        foreach ($surveys as $survey) {
-            if ($survey->name == 'myChangedSurvey') {
-                $s = $survey;
-                break;
-            }
-        }
-        $this->assertNotNull($s);
-        $this->assertEquals($s->name, 'myChangedSurvey');
-        $this->assertEquals($s->url, 'http://survey.example/changed');
-        $this->assertEquals($s->active, false);
-        $this->assertEquals($s->target, 'views["editor"].ratio > 0.5');
-    }
-
-    public function testSurveyDelete()
-    {
-        $json = '{
-            "name": "myInactiveSurvey",
-            "url": "http://survey.example/inactive",
-            "active": false,
-            "uuid": "{bdfa82c5-238f-404b-a441-07ca3d3eff7f}"
-        }';
-
-        $s = Survey::fromJson($json);
-        $s->delete(self::$db);
-
-        $surveys = Survey::surveysForProduct(self::$db, 'org.kde.UnitTest');
-        $s = null;
-        foreach ($surveys as $survey) {
-            if ($survey->name == 'myInactiveSurvey') {
-                $s = $survey;
-                break;
-            }
-        }
-        $this->assertNull($s);
-    }
-
-    public static function invalidInput_data()
-    {
-        return [
-            'empty' => [ '{}' ],
-            'mising url' => [ '{ "name": "surveyName" }' ],
-            'missing active' => [ '{ "name": "surveyName", "url": "http://survey.example/" }' ],
-            'empty name' => [ '{ "name": "", "url": "http://survey.example/", "active": false }' ],
-            'non-string name' => [ '{ "name": 123, "url": "http://survey.example/", "active": false }' ],
-            'non-numberic id' => [ '{ "name": "myName", "url": "http://survey.example/", "active": true, "id": "string" }' ]
-        ];
-    }
-
-    /**
-     * @dataProvider invalidInput_data
-     */
-    public function testInvalidInput($json)
-    {
-        $this->expectException(RESTException::class);
-        $this->expectExceptionCode(400);
-        $s = Survey::fromJson($json);
-    }
-}
-
-
+<?php
+/*
+    SPDX-FileCopyrightText: 2016 Volker Krause <vkrause@kde.org>
+
+    SPDX-License-Identifier: MIT
+*/
+
+require_once('datastoretesthelper.php');
+require_once('../src/server/shared/product.php');
+require_once('../src/server/shared/survey.php');
+
+class SurveyTest extends PHPUnit\Framework\TestCase
+{
+    private static $db;
+
+    public static function setUpBeforeClass(): void
+    {
+        self::$db = DatastoreTestHelper::setup();
+    }
+
+    public function testFromJson()
+    {
+        $s = Survey::fromJson('{
+            "name": "mySurvey",
+            "url": "http://survey.example/",
+            "active": true,
+            "target": "screen[0].dpi > 100",
+            "uuid": "{9e529dfa-0213-413e-a1a8-8a9cea7d5a97}"
+        }');
+        $this->assertEquals($s->name, 'mySurvey');
+        $this->assertEquals($s->url, 'http://survey.example/');
+        $this->assertEquals($s->active, true);
+        $this->assertEquals($s->uuid, '{9e529dfa-0213-413e-a1a8-8a9cea7d5a97}');
+        $this->assertEquals($s->target, 'screen[0].dpi > 100');
+    }
+
+    public function testToJson()
+    {
+        $s = new Survey();
+        $s->name = 'mySurvey';
+        $s->url = 'http://survey.example/';
+        $s->active = true;
+        $s->uuid = '{9e529dfa-0213-413e-a1a8-8a9cea7d5a97}';
+        $s->target = 'screen[0].dpi > 100';
+
+        $this->assertJsonStringEqualsJsonString(json_encode($s), '{
+            "uuid": "{9e529dfa-0213-413e-a1a8-8a9cea7d5a97}",
+            "name": "mySurvey",
+            "url": "http://survey.example/",
+            "active": true,
+            "target": "screen[0].dpi > 100"
+        }');
+    }
+
+    public function testGetSurveys()
+    {
+        $surveys = Survey::surveysForProduct(self::$db, 'I don\'t exist');
+        $this->assertCount(0, $surveys);
+
+        $surveys = Survey::surveysForProduct(self::$db, 'org.kde.UnitTest');
+        $this->assertCount(2, $surveys);
+
+        $survey1 = $surveys[0];
+        $this->assertInstanceOf(Survey::class, $survey1);
+        $survey2 = $surveys[1];
+        $this->assertInstanceOf(Survey::class, $survey2);
+
+        if ($survey1->name > $survey2->name) {
+            $survey1 = $surveys->schema[1];
+            $survey2 = $surveys->schema[0];
+        }
+
+        $this->assertEquals($survey1->name, 'myActiveSurvey');
+        $this->assertEquals($survey1->url, 'http://survey.example/active');
+        $this->assertEquals($survey1->active, true);
+        $this->assertEquals($survey1->target, 'usageTime.value >= 3600');
+
+        $this->assertEquals($survey2->name, 'myInactiveSurvey');
+        $this->assertEquals($survey2->url, 'http://survey.example/inactive');
+        $this->assertEquals($survey2->active, false);
+        $this->assertEquals($survey2->target, 'screen[0].dpi < 200');
+
+        $p = Product::productByName(self::$db, 'org.kde.UnitTest');
+        $this->assertNotNull($p);
+        $surveys = Survey::activeSurveysForProduct(self::$db, $p);
+        $this->assertCount(1, $surveys);
+
+        $s = $surveys[0];
+        $this->assertInstanceOf(Survey::class, $s);
+        $this->assertEquals($s->name, 'myActiveSurvey');
+        $this->assertEquals($s->url, 'http://survey.example/active');
+        $this->assertEquals($s->active, true);
+        $this->assertEquals($s->target, 'usageTime.value >= 3600');
+    }
+
+    public function testSurveyInsert()
+    {
+        $json = '{
+            "name": "newSurvey",
+            "uuid": "{9e529dfa-0213-413e-a1a8-8a9cea7d5a97}",
+            "url": "http://survey.example/new",
+            "active": true,
+            "target": "startCount.value > 10"
+        }';
+
+        $p = Product::productByName(self::$db, 'org.kde.UnitTest');
+        $this->assertNotNull($p);
+
+        $s = Survey::fromJson($json);
+        $s->insert(self::$db, $p);
+
+        $surveys = Survey::activeSurveysForProduct(self::$db, $p);
+        $s = null;
+        foreach ($surveys as $survey) {
+            if ($survey->name == 'newSurvey') {
+                $s = $survey;
+                break;
+            }
+        }
+        $this->assertNotNull($s);
+        $this->assertEquals($s->name, 'newSurvey');
+        $this->assertEquals($s->url, 'http://survey.example/new');
+        $this->assertEquals($s->active, true);
+        $this->assertEquals($s->target, 'startCount.value > 10');
+        $this->assertEquals($s->uuid, '{9e529dfa-0213-413e-a1a8-8a9cea7d5a97}');
+    }
+
+    public function testSurveyUpdate()
+    {
+        $json = '{
+            "name": "myChangedSurvey",
+            "url": "http://survey.example/changed",
+            "active": false,
+            "uuid": "{962bbd80-7120-4f18-a4c0-5800fa323868}",
+            "target": "views[\"editor\"].ratio > 0.5"
+        }';
+
+        $s = Survey::fromJson($json);
+        $s->update(self::$db);
+
+        $surveys = Survey::surveysForProduct(self::$db, 'org.kde.UnitTest');
+        $s = null;
+        foreach ($surveys as $survey) {
+            if ($survey->name == 'myChangedSurvey') {
+                $s = $survey;
+                break;
+            }
+        }
+        $this->assertNotNull($s);
+        $this->assertEquals($s->name, 'myChangedSurvey');
+        $this->assertEquals($s->url, 'http://survey.example/changed');
+        $this->assertEquals($s->active, false);
+        $this->assertEquals($s->target, 'views["editor"].ratio > 0.5');
+    }
+
+    public function testSurveyDelete()
+    {
+        $json = '{
+            "name": "myInactiveSurvey",
+            "url": "http://survey.example/inactive",
+            "active": false,
+            "uuid": "{bdfa82c5-238f-404b-a441-07ca3d3eff7f}"
+        }';
+
+        $s = Survey::fromJson($json);
+        $s->delete(self::$db);
+
+        $surveys = Survey::surveysForProduct(self::$db, 'org.kde.UnitTest');
+        $s = null;
+        foreach ($surveys as $survey) {
+            if ($survey->name == 'myInactiveSurvey') {
+                $s = $survey;
+                break;
+            }
+        }
+        $this->assertNull($s);
+    }
+
+    public static function invalidInput_data()
+    {
+        return [
+            'empty' => [ '{}' ],
+            'mising url' => [ '{ "name": "surveyName" }' ],
+            'missing active' => [ '{ "name": "surveyName", "url": "http://survey.example/" }' ],
+            'empty name' => [ '{ "name": "", "url": "http://survey.example/", "active": false }' ],
+            'non-string name' => [ '{ "name": 123, "url": "http://survey.example/", "active": false }' ],
+            'non-numberic id' => [ '{ "name": "myName", "url": "http://survey.example/", "active": true, "id": "string" }' ]
+        ];
+    }
+
+    #[\PHPUnit\Framework\Attributes\DataProvider('invalidInput_data')]
+    public function testInvalidInput($json)
+    {
+        $this->expectException(RESTException::class);
+        $this->expectExceptionCode(400);
+        $s = Survey::fromJson($json);
+    }
+}
+
+
diff --git a/autotests/aggregationtest.php b/autotests/aggregationtest.php
--- a/autotests/aggregationtest.php
+++ b/autotests/aggregationtest.php
@@ -1,121 +1,119 @@
-<?php
-/*
-    SPDX-FileCopyrightText: 2016 Volker Krause <vkrause@kde.org>
-
-    SPDX-License-Identifier: MIT
-*/
-
-require_once('datastoretesthelper.php');
-require_once('../src/server/shared/aggregation.php');
-
-class AggregationTest extends PHPUnit\Framework\TestCase
-{
-    private static $db;
-
-    public static function setUpBeforeClass(): void
-    {
-        self::$db = DatastoreTestHelper::setup();
-    }
-
-    public function testFromJson()
-    {
-        $json = '[]';
-        $aggrs = Aggregation::fromJson(json_decode($json));
-        $this->assertCount(0, $aggrs);
-
-        $json = '[
-            { "type": "category", "name": "category elem11", "elements": [ { "type": "value", "schemaEntry": "entry1", "schemaEntryElement": "elem11" } ] },
-            { "type": "numeric", "name": "size: entry2", "elements": [ { "type": "size", "schemaEntry": "entry2" } ] }
-        ]';
-        $aggrs = Aggregation::fromJson(json_decode($json));
-        $this->assertCount(2, $aggrs);
-
-        $a = $aggrs[0];
-        $this->assertInstanceOf(Aggregation::class, $a);
-        $this->assertEquals('category', $a->type);
-        $this->assertEquals('category elem11', $a->name);
-        $this->assertCount(1, $a->elements);
-        $a = $aggrs[1];
-        $this->assertInstanceOf(Aggregation::class, $a);
-        $this->assertEquals('numeric', $a->type);
-        $this->assertEquals('size: entry2', $a->name);
-        $this->assertCount(1, $a->elements);
-    }
-
-    public function testLoad()
-    {
-        $p = Product::productByName(self::$db, 'org.kde.UnitTest');
-        $this->assertNotNull($p);
-
-        $aggrs = Aggregation::aggregationsForProduct(self::$db, $p);
-        $this->assertCount(2, $aggrs);
-
-        $a = $aggrs[0];
-        $this->assertInstanceOf(Aggregation::class, $a);
-        $this->assertEquals('category', $a->type);
-        $this->assertCount(1, $a->elements);
-
-        $a = $aggrs[1];
-        $this->assertInstanceOf(Aggregation::class, $a);
-        $this->assertEquals('numeric', $a->type);
-        $this->assertCount(1, $a->elements);
-    }
-
-    public function testWrite()
-    {
-        $p = Product::productByName(self::$db, 'org.kde.UnitTest');
-        $this->assertNotNull($p);
-
-        $a = new Aggregation;
-        $a->type = 'xy';
-        $a->name = 'n1';
-        $a->elements = json_decode('[
-            { "type": "value", "schemaEntry": "entry2", "schemaEntryElement": "element21" },
-            { "type": "value", "schemaEntry": "entry2", "schemaEntryElement": "element22" }
-        ]');
-        Aggregation::update(self::$db, $p, array(0 => $a));
-
-        $aggrs = Aggregation::aggregationsForProduct(self::$db, $p);
-        $this->assertCount(1, $aggrs);
-
-        $a = $aggrs[0];
-        $this->assertInstanceOf(Aggregation::class, $a);
-        $this->assertEquals('xy', $a->type);
-        $this->assertEquals('n1', $a->name);
-        $this->assertCount(2, $a->elements);
-    }
-
-    public function testDelete()
-    {
-        $p = Product::productByName(self::$db, 'org.kde.UnitTest');
-        $this->assertNotNull($p);
-
-        $aggrs = Aggregation::aggregationsForProduct(self::$db, $p);
-        $this->assertNotEmpty($aggrs);
-
-        Aggregation::delete(self::$db, $p);
-        $aggrs = Aggregation::aggregationsForProduct(self::$db, $p);
-        $this->assertEmpty($aggrs);
-    }
-
-    public static function invalidJson_data()
-    {
-        return [
-            'nothing' => [ '' ],
-            'object' => [ '{}' ],
-            'array of non-objects' => [ '[1, 2, 3]' ],
-            'missing type' => [ '[{ "elements": [] }]' ],
-            'missing name' => [ '[{ "type": "category", "elements": [] }]' ]
-        ];
-    }
-
-    /**
-     * @dataProvider invalidJson_data
-     */
-    public function testInvalidJson($input)
-    {
-        $this->expectException(RESTException::class);
-        $this->expectExceptionCode(400);
-        $aggrs = Aggregation::fromJson(json_decode($input));
-    }
-}
+<?php
+/*
+    SPDX-FileCopyrightText: 2016 Volker Krause <vkrause@kde.org>
+
+    SPDX-License-Identifier: MIT
+*/
+
+require_once('datastoretesthelper.php');
+require_once('../src/server/shared/aggregation.php');
+
+class AggregationTest extends PHPUnit\Framework\TestCase
+{
+    private static $db;
+
+    public static function setUpBeforeClass(): void
+    {
+        self::$db = DatastoreTestHelper::setup();
+    }
+
+    public function testFromJson()
+    {
+        $json = '[]';
+        $aggrs = Aggregation::fromJson(json_decode($json));
+        $this->assertCount(0, $aggrs);
+
+        $json = '[
+            { "type": "category", "name": "category elem11", "elements": [ { "type": "value", "schemaEntry": "entry1", "schemaEntryElement": "elem11" } ] },
+            { "type": "numeric", "name": "size: entry2", "elements": [ { "type": "size", "schemaEntry": "entry2" } ] }
+        ]';
+        $aggrs = Aggregation::fromJson(json_decode($json));
+        $this->assertCount(2, $aggrs);
+
+        $a = $aggrs[0];
+        $this->assertInstanceOf(Aggregation::class, $a);
+        $this->assertEquals('category', $a->type);
+        $this->assertEquals('category elem11', $a->name);
+        $this->assertCount(1, $a->elements);
+        $a = $aggrs[1];
+        $this->assertInstanceOf(Aggregation::class, $a);
+        $this->assertEquals('numeric', $a->type);
+        $this->assertEquals('size: entry2', $a->name);
+        $this->assertCount(1, $a->elements);
+    }
+
+    public function testLoad()
+    {
+        $p = Product::productByName(self::$db, 'org.kde.UnitTest');
+        $this->assertNotNull($p);
+
+        $aggrs = Aggregation::aggregationsForProduct(self::$db, $p);
+        $this->assertCount(2, $aggrs);
+
+        $a = $aggrs[0];
+        $this->assertInstanceOf(Aggregation::class, $a);
+        $this->assertEquals('category', $a->type);
+        $this->assertCount(1, $a->elements);
+
+        $a = $aggrs[1];
+        $this->assertInstanceOf(Aggregation::class, $a);
+        $this->assertEquals('numeric', $a->type);
+        $this->assertCount(1, $a->elements);
+    }
+
+    public function testWrite()
+    {
+        $p = Product::productByName(self::$db, 'org.kde.UnitTest');
+        $this->assertNotNull($p);
+
+        $a = new Aggregation;
+        $a->type = 'xy';
+        $a->name = 'n1';
+        $a->elements = json_decode('[
+            { "type": "value", "schemaEntry": "entry2", "schemaEntryElement": "element21" },
+            { "type": "value", "schemaEntry": "entry2", "schemaEntryElement": "element22" }
+        ]');
+        Aggregation::update(self::$db, $p, array(0 => $a));
+
+        $aggrs = Aggregation::aggregationsForProduct(self::$db, $p);
+        $this->assertCount(1, $aggrs);
+
+        $a = $aggrs[0];
+        $this->assertInstanceOf(Aggregation::class, $a);
+        $this->assertEquals('xy', $a->type);
+        $this->assertEquals('n1', $a->name);
+        $this->assertCount(2, $a->elements);
+    }
+
+    public function testDelete()
+    {
+        $p = Product::productByName(self::$db, 'org.kde.UnitTest');
+        $this->assertNotNull($p);
+
+        $aggrs = Aggregation::aggregationsForProduct(self::$db, $p);
+        $this->assertNotEmpty($aggrs);
+
+        Aggregation::delete(self::$db, $p);
+        $aggrs = Aggregation::aggregationsForProduct(self::$db, $p);
+        $this->assertEmpty($aggrs);
+    }
+
+    public static function invalidJson_data()
+    {
+        return [
+            'nothing' => [ '' ],
+            'object' => [ '{}' ],
+            'array of non-objects' => [ '[1, 2, 3]' ],
+            'missing type' => [ '[{ "elements": [] }]' ],
+            'missing name' => [ '[{ "type": "category", "elements": [] }]' ]
+        ];
+    }
+
+    #[\PHPUnit\Framework\Attributes\DataProvider('invalidJson_data')]
+    public function testInvalidJson($input)
+    {
+        $this->expectException(RESTException::class);
+        $this->expectExceptionCode(400);
+        $aggrs = Aggregation::fromJson(json_decode($input));
+    }
+}
diff --git a/autotests/datastoretesthelper.php b/autotests/datastoretesthelper.php
--- a/autotests/datastoretesthelper.php
+++ b/autotests/datastoretesthelper.php
@@ -1,43 +1,43 @@
-<?php
-/*
-    SPDX-FileCopyrightText: 2017-2023 Volker Krause <vkrause@kde.org>
-    SPDX-License-Identifier: MIT
-*/
-
-require_once('../src/server/shared/datastore.php');
-
-// default database configuration for sqlite, can be overridden by localconfig.php
-$USERFEEDBACK_DB_DRIVER = 'sqlite';
-$USERFEEDBACK_DB_NAME =  ':memory:';
-
-class DatastoreTestHelper
-{
-    public static function setup()
-    {
-        // database schema setup
-        $db = new Datastore();
-        $db->checkSchema();
-
-        // add dummy test data (from previous use of DBUnit, might make sense to move this to the corresponding tests relying on this instead)
-        $db->pdoHandle()->exec('INSERT INTO tbl_product (col_id, col_name) VALUES (2, "org.kde.UnitTest")');
-
-        $db->pdoHandle()->exec('INSERT INTO tbl_schema (col_id, col_product_id, col_name, col_type) VALUES (31, 2, "entry1", "scalar")');
-        $db->pdoHandle()->exec('INSERT INTO tbl_schema (col_id, col_product_id, col_name, col_type) VALUES (32, 2, "entry2", "list")');
-
-        $db->pdoHandle()->exec('INSERT INTO tbl_schema_element (col_id, col_schema_id, col_name, col_type) VALUES (42, 31, "element11", "string")');
-        $db->pdoHandle()->exec('INSERT INTO tbl_schema_element (col_id, col_schema_id, col_name, col_type) VALUES (43, 31, "element12", "bool")');
-        $db->pdoHandle()->exec('INSERT INTO tbl_schema_element (col_id, col_schema_id, col_name, col_type) VALUES (44, 32, "element21", "int")');
-        $db->pdoHandle()->exec('INSERT INTO tbl_schema_element (col_id, col_schema_id, col_name, col_type) VALUES (45, 32, "element22", "number")');
-
-        $db->pdoHandle()->exec('INSERT INTO tbl_survey (col_id, col_uuid, col_product_id, col_name, col_url, col_active, col_target) VALUES (101, "{962bbd80-7120-4f18-a4c0-5800fa323868}", 2, "myActiveSurvey", "http://survey.example/active", 1, "usageTime.value >= 3600")');
-        $db->pdoHandle()->exec('INSERT INTO tbl_survey (col_id, col_uuid, col_product_id, col_name, col_url, col_active, col_target) VALUES (102, "{bdfa82c5-238f-404b-a441-07ca3d3eff7f}", 2, "myInactiveSurvey", "http://survey.example/inactive", 0, "screen[0].dpi < 200")');
-
-        $stmt =
-        $db->pdoHandle()->exec("INSERT INTO tbl_aggregation (col_id, col_name, col_product_id, col_type, col_elements) VALUES (201, 'Category: entry1.elem11', 2, 'category', '[ { \"type\": \"value\", \"schemaEntry\": \"entry1\", \"schemaEntryElement\": \"element11\" } ]')");
-        $db->pdoHandle()->exec("INSERT INTO tbl_aggregation (col_id, col_name, col_product_id, col_type, col_elements) VALUES (202, 'entry2 size distribution', 2, 'numeric', '[ { \"type\": \"size\", \"schemaEntry\": \"entry2\" } ]')");
-
-        return $db;
-    }
-}
-
-?>
+<?php
+/*
+    SPDX-FileCopyrightText: 2017-2023 Volker Krause <vkrause@kde.org>
+    SPDX-License-Identifier: MIT
+*/
+
+require_once('../src/server/shared/datastore.php');
+
+// default database configuration for sqlite, can be overridden by localconfig.php
+$GLOBALS['USERFEEDBACK_DB_DRIVER'] = 'sqlite';
+$GLOBALS['USERFEEDBACK_DB_NAME'] = ':memory:';
+
+class DatastoreTestHelper
+{
+    public static function setup()
+    {
+        // database schema setup
+        $db = new Datastore();
+        $db->checkSchema();
+
+        // add dummy test data (from previous use of DBUnit, might make sense to move this to the corresponding tests relying on this instead)
+        $db->pdoHandle()->exec('INSERT INTO tbl_product (col_id, col_name) VALUES (2, "org.kde.UnitTest")');
+
+        $db->pdoHandle()->exec('INSERT INTO tbl_schema (col_id, col_product_id, col_name, col_type) VALUES (31, 2, "entry1", "scalar")');
+        $db->pdoHandle()->exec('INSERT INTO tbl_schema (col_id, col_product_id, col_name, col_type) VALUES (32, 2, "entry2", "list")');
+
+        $db->pdoHandle()->exec('INSERT INTO tbl_schema_element (col_id, col_schema_id, col_name, col_type) VALUES (42, 31, "element11", "string")');
+        $db->pdoHandle()->exec('INSERT INTO tbl_schema_element (col_id, col_schema_id, col_name, col_type) VALUES (43, 31, "element12", "bool")');
+        $db->pdoHandle()->exec('INSERT INTO tbl_schema_element (col_id, col_schema_id, col_name, col_type) VALUES (44, 32, "element21", "int")');
+        $db->pdoHandle()->exec('INSERT INTO tbl_schema_element (col_id, col_schema_id, col_name, col_type) VALUES (45, 32, "element22", "number")');
+
+        $db->pdoHandle()->exec('INSERT INTO tbl_survey (col_id, col_uuid, col_product_id, col_name, col_url, col_active, col_target) VALUES (101, "{962bbd80-7120-4f18-a4c0-5800fa323868}", 2, "myActiveSurvey", "http://survey.example/active", 1, "usageTime.value >= 3600")');
+        $db->pdoHandle()->exec('INSERT INTO tbl_survey (col_id, col_uuid, col_product_id, col_name, col_url, col_active, col_target) VALUES (102, "{bdfa82c5-238f-404b-a441-07ca3d3eff7f}", 2, "myInactiveSurvey", "http://survey.example/inactive", 0, "screen[0].dpi < 200")');
+
+        $stmt =
+        $db->pdoHandle()->exec("INSERT INTO tbl_aggregation (col_id, col_name, col_product_id, col_type, col_elements) VALUES (201, 'Category: entry1.elem11', 2, 'category', '[ { \"type\": \"value\", \"schemaEntry\": \"entry1\", \"schemaEntryElement\": \"element11\" } ]')");
+        $db->pdoHandle()->exec("INSERT INTO tbl_aggregation (col_id, col_name, col_product_id, col_type, col_elements) VALUES (202, 'entry2 size distribution', 2, 'numeric', '[ { \"type\": \"size\", \"schemaEntry\": \"entry2\" } ]')");
+
+        return $db;
+    }
+}
+
+?>
